Tracking/src/utils.cpp

```diff
@@ -0,0 +1,356 @@
+#include "utils.hpp"
+
+dd4hep::rec::LayeredCalorimeterData * getExtension(unsigned int includeFlag, unsigned int excludeFlag) {
+
+    dd4hep::rec::LayeredCalorimeterData * theExtension = 0;
+    dd4hep::Detector& detector = dd4hep::Detector::getInstance();
+    if (detector.detectors().empty()) {
+        throw std::runtime_error("Detector is not initialized: no DetElements found.");
+    }
+
+    const std::vector< dd4hep::DetElement>& theDetectors = dd4hep::DetectorSelector(detector).detectors(  includeFlag, excludeFlag );
+
+    int debug_lvl = 0; 
+    if (debug_lvl > 0) {
+         std::cout << " getExtension :  includeFlag: " << dd4hep::DetType( includeFlag ) << " excludeFlag: " << dd4hep::DetType( excludeFlag )
+            << "  found : " << theDetectors.size() << "  - first det: " << theDetectors.at(0).name()
+            << std::endl;
+        
+    }
+       
+    if( theDetectors.size()  != 1 ){
+
+      std::stringstream es ;
+      es << " getExtension: selection is not unique (or empty)  includeFlag: " << dd4hep::DetType( includeFlag ) << " excludeFlag: " << dd4hep::DetType( excludeFlag )
+        << " --- found detectors : " ;
+      for( unsigned i=0, N= theDetectors.size(); i<N ; ++i ){
+        es << theDetectors.at(i).name() << ", " ;
+      }
+      throw std::runtime_error( es.str() ) ;
+    }
+
+    theExtension = theDetectors.at(0).extension<dd4hep::rec::LayeredCalorimeterData>();
+
+    return theExtension;
+}
+
+edm4hep::TrackState getExtrapolationAtCalorimeter(const pandora::CartesianVector& ecalProjection, const HelixClass_double& helixAtLastHit,double Bz) {
```

Comment:
As far as I understand, the extrapolation function still only works for ALLEGRO, but not IDEA. So we do still need #62 to retrieve the right parameter, right?